feat: Introduce biome

volnei opened the pull request:
## What does this PR do?

Replaces ESLint and Prettier with [Biome](https://biomejs.dev/) across the monorepo to unify linting and formatting, simplify config, and dramatically improve performance.

## Benefits

### Performance Improvements

| Operation | ESLint/Prettier (main) | Biome (this branch) | Improvement |
|-----------|------------------------|---------------------|-------------|
| **Lint** | 1m 55s | 19s | **~6x faster** |
| **Format check** | 3m 57s | 6s | **~40x faster** |

*Benchmarks run on the full monorepo (~8000 files)*

### Developer Experience
- **Single tool**: One unified tool for both linting and formatting instead of managing ESLint + Prettier + plugins
- **Faster CI**: Significantly reduced CI pipeline times for lint and format checks
- **Faster local development**: Near-instant feedback on lint/format issues
- **Simpler configuration**: Single `biome.json` file instead of multiple ESLint/Prettier configs
- **Better VSCode integration**: Native Biome extension with format-on-save and auto-fix

### Reduced Dependencies

| Category | Removed |
|----------|---------|
| **Config files deleted** | 36 files |
| **npm packages removed** | 18 packages |
| **Internal packages removed** | 2 packages |

**Removed npm packages:**
- `eslint` and 11 ESLint plugins (`eslint-plugin-react`, `eslint-plugin-react-hooks`, `eslint-plugin-import`, `eslint-plugin-turbo`, `eslint-plugin-playwright`, `eslint-plugin-prettier`, `eslint-plugin-only-warn`, `eslint-config-next`, `eslint-config-prettier`, `eslint-config-turbo`, `eslint-import-resolver-typescript`)
- `@typescript-eslint/eslint-plugin`, `@typescript-eslint/parser`, `@typescript-eslint/utils`
- `prettier`, `prettier-plugin-tailwindcss`, `@trivago/prettier-plugin-sort-imports`

**Removed internal packages:**
- `@calcom/eslint-config`
- `@calcom/eslint-plugin`

## Key Changes

- Added `biome.json` at repo root with formatter settings matching previous Prettier config (lineWidth: 110, semicolons, trailing commas, etc.)
- Ported React Hooks linting rules: `useHookAtTopLevel` (error), `useExhaustiveDependencies` (warn)
- Ported all restricted import rules from ESLint configs to Biome overrides:
  - `packages/lib`: blocks imports from `../app-store/**`, `../features/**`, `../trpc/**`, and `@trpc/server` (with exceptions for `defaultResponder.ts`)
  - `packages/features`: blocks imports from `../trpc/**`
  - `packages/app-store`: blocks imports from `../features/**`, `../trpc/**`
  - `packages/trpc`: blocks imports from `../apps/web/**` (excludes `types/**`)
- Updated all package.json scripts to use `biome lint` and `biome format`
- Removed ESLint/Prettier configs and internal ESLint plugin packages
- Updated VSCode settings and extension recommendations to use Biome
- Enabled `tailwindDirectives` in Biome CSS parser for Tailwind CSS support

## Additional Fixes

- Fixed TypeScript error in `apps/api/v2/src/swagger/generate-swagger.ts`: renamed `const require` to `const nodeRequire` to avoid conflict with TypeScript's reserved `require` identifier
- Replaced ESLint usage in `apps/web/playwright/embed-code-generator.e2e.ts` with Biome CLI for validating embed code snippets
- Fixed unhandled error in `packages/embeds/embed-core/src/embed-iframe.ts`: Added guard in `makeBodyVisible()` to check if `document` exists before accessing it, preventing errors when timers fire after test environment teardown
- Added explicit `stripe: "9.16.0"` dependency to `packages/features` to stabilize type resolution after ESLint/Prettier removal changed Yarn's hoisting behavior

## Updates Since Last Revision

Addressed cubic review comments:
- Fixed invalid `--reporter-path` Biome CLI option by using shell redirection (`biome lint --reporter json . > file.json`) in apps/web, packages/lib, packages/app-store, and packages/ui
- Fixed packages/lib lint report filename from `app-store.json` to `lib.json`
- Added `eslint`, `eslint-plugin-react-compiler`, and `typescript-eslint` back to companion/package.json (required for `lint:react-compiler` script which still uses ESLint)
- Regenerated companion/bun.lock after adding ESLint dependencies

## Known Incompatibilities (for follow-up PRs)

- Import ordering (`@trivago/prettier-plugin-sort-imports`) - Biome's import organization differs
- Tailwind class sorting (`prettier-plugin-tailwindcss`) - No Biome equivalent
- Next.js specific rules - No Biome equivalent
- `turbo/no-undeclared-env-vars` - No Biome equivalent

## Human Review Checklist

- [ ] Verify restricted import rules in `biome.json` match the original ESLint configs
- [ ] Confirm companion package's ESLint retention for `lint:react-compiler` is acceptable
- [ ] Verify `stripe` dependency in packages/features is needed (fixes hoisting issue causing type errors)
- [ ] Test `lint:report` scripts work correctly with shell redirection

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] I have updated the developer docs in /docs if this PR makes changes that would require a documentation change. Updated `agents/commands.md` with new lint/format commands.
- [x] I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?

1. Run `yarn lint` - should use Biome and pass
2. Run `yarn format` - should format with Biome
3. Verify restricted imports are enforced by creating a test file in `packages/lib` that imports from `@trpc/server` - should error
4. Verify `packages/lib/server/defaultResponder.ts` is exempt from the `@trpc/server` restriction
5. Run `yarn lint:report` in apps/web - should generate JSON report via shell redirection

---

**Link to Devin run:** https://app.devin.ai/sessions/0c0d90a0e4cb4524a9243f4b80701bc6
**Requested by:** Volnei Munhoz (@volnei)

volnei committed:
Configure biome

volnei commented:
@cubic-dev-ai Please create a comprehensive list on what's changed by introducing biome and all important information for developers to compose a document to be shared with engineers

keithwillcode commented on apps/api/v2/src/modules/stripe/utils/newStripeInstance.ts:
```diff
@@ -2,5 +2,5 @@ import Stripe from "stripe";
 const stripeApiKey = process.env.STRIPE_API_KEY || "";
 export const stripeInstance = new Stripe(stripeApiKey, {
-  apiVersion: "2020-08-27",
+  apiVersion: "2024-04-10",
```
@volnei I don't think we can easily do this, can we?

keithwillcode commented on apps/api/v2/src/modules/organizations/stripe/services/organizations-stripe.service.ts:
```diff
@@ -46,11 +46,11 @@ export class OrganizationsStripeService {
-      const account = await stripeInstance.accounts.retrieve(response["stripe_user_id"]);
-      data["default_currency"] = account.default_currency;
-    }
+      const data: StripeData = { ...response, default_currency: "" };
```
@volnei Are these extra spaces correct? Seems too indented

volnei replied:
old commit... indentation was wrong on main.

keithwillcode commented on packages/app-store/_utils/stripe.ts:
```diff
@@ -68,5 +68,5 @@ export async function getStripeCustomerId(user: UserType): Promise<string> {
 const stripePrivateKey = process.env.STRIPE_PRIVATE_KEY || "";
 export const stripe = new Stripe(stripePrivateKey, {
-  apiVersion: "2020-08-27",
+  apiVersion: "2024-04-10",
```
This should not be upgraded

keithwillcode commented on packages/app-store/analytics.services.generated.ts:
```diff
@@ -2,9 +2,6 @@
-        dub: import("./dub/lib/AnalyticsService"),
-      };
+export const AnalyticsServiceMap = process.env.NEXT_PUBLIC_IS_E2E === '1' ? {} : {
+"dub": import("./dub/lib/AnalyticsService"),
```
this formatting is weird. no indention here?

keithwillcode commented on packages/app-store/apps.browser.generated.tsx:
```diff
@@ -2,70 +2,61 @@
-  exchange2016calendar: dynamic(() => import("./exchange2016calendar/components/InstallAppButton")),
-  office365video: dynamic(() => import("./office365video/components/InstallAppButton")),
-  vital: dynamic(() => import("./vital/components/InstallAppButton")),
+"exchange2013calendar": dynamic(() => import("./exchange2013calendar/components/InstallAppButton")),
```
this formatting is weird. no indention here?

keithwillcode commented on packages/app-store/apps.browser.generated.tsx:
```diff
@@ -2,70 +2,61 @@
-  qr_code: dynamic(() => import("./qr_code/components/EventTypeAppSettingsInterface")),
-  stripepayment: dynamic(() => import("./stripepayment/components/EventTypeAppSettingsInterface")),
-};
+"alby": dynamic(() => import("./alby/components/EventTypeAppSettingsInterface")),
```
this formatting is weird. no indention here?

keithwillcode commented on packages/app-store/apps.schemas.generated.ts:
```diff
@@ -2,106 +2,105 @@
-  zohocrm: zohocrm_zod_ts,
-  zoomvideo: zoomvideo_zod_ts,
-};
+"alby": alby_zod_ts,
```
this formatting is weird. no indention here?

keithwillcode commented on packages/app-store/bookerApps.metadata.generated.ts:
```diff
@@ -2,94 +2,93 @@
-import whereby_config_json from "./whereby/config.json";
-import { metadata as zoomvideo__metadata_ts } from "./zoomvideo/_metadata";
-
+import campfire_config_json from "./campfire/config.json"
```
this formatting is weird. no indention here?

devin-ai-integration[bot] committed:
Merge origin/main into feat/introduce-biome - resolve conflicts

Co-Authored-By: Volnei Munhoz <volnei.munhoz@gmail.com>

volnei committed:
Fix companion build

volnei committed:
Merge branch 'main' into feat/introduce-biome

devin-ai-integration[bot] committed:
fix: remove generated files from formatter ignore list to enable proper formatting

Co-Authored-By: Volnei Munhoz <volnei.munhoz@gmail.com>

volnei committed:
Merge branch 'main' into feat/introduce-biome

devin-ai-integration[bot] committed:
fix: add explicit stripe dependency to @calcom/features to fix type resolution

Co-Authored-By: Volnei Munhoz <volnei.munhoz@gmail.com>

devin-ai-integration[bot] committed:
Merge branch 'feat/introduce-biome' of https://git-manager.devin.ai/proxy/github.com/calcom/cal.com into feat/introduce-biome

devin-ai-integration[bot] committed:
fix: rename const require to nodeRequire in generate-swagger.ts to avoid TypeScript reserved identifier conflict

Co-Authored-By: Volnei Munhoz <volnei.munhoz@gmail.com>

devin-ai-integration[bot] committed:
fix: add guard for document in makeBodyVisible to prevent test environment teardown errors

Co-Authored-By: Volnei Munhoz <volnei.munhoz@gmail.com>

devin-ai-integration[bot] committed:
fix: replace ESLint with Biome CLI in embed-code-generator.e2e.ts

Co-Authored-By: Volnei Munhoz <volnei.munhoz@gmail.com>

devin-ai-integration[bot] committed:
fix: address cubic review comments

- Fix invalid --reporter-path Biome CLI option by using shell redirection
- Fix packages/lib lint report filename (app-store.json -> lib.json)
- Add typescript-eslint and eslint back to companion for lint:react-compiler
- Add missing restricted import rules to biome.json:
  - packages/lib: add ../trpc/** and @trpc/server restrictions
  - packages/trpc: add ../apps/web/** restriction

Co-Authored-By: Volnei Munhoz <volnei.munhoz@gmail.com>

devin-ai-integration[bot] committed:
chore: regenerate companion bun.lock after adding eslint dependencies

Co-Authored-By: Volnei Munhoz <volnei.munhoz@gmail.com>

volnei commented on packages/features/package.json:
```diff
@@ -30,6 +30,7 @@
     "react-select": "5.8.0",
     "react-sticky-box": "2.0.4",
     "recharts": "3.0.2",
+    "stripe": "9.16.0",
```
Necessary due to hoist flakiness

volnei commented on packages/app-store/apps.browser.generated.tsx:
```diff
@@ -3,19 +3,18 @@
     Don't modify this file manually.
 **/
 import dynamic from "next/dynamic";
-
```
the file is not being generated with new line... acceptable since biome keeps the format

volnei commented on companion/package.json:
```diff
@@ -87,20 +87,22 @@
+      "babel-plugin-react-compiler": "19.0.0-beta-af1b7da-20250417",
+      "babel-preset-expo": "54.0.7",
+      "eslint": "9.28.0",
+      "eslint-plugin-react-compiler": "19.0.0-beta-af1b7da-20250417",
```
companion needs eslint due to it's compiler

pedroccastro reviewed (changes requested):
Hey @volnei ! Tested locally and noticed `packages/coss-ui` might have been missed in the migration, it's still referencing eslint and missing the biome dependency. Also, I believe there's a missing `css` section in biome.json (left an inline comment)

volnei committed:
Remove remaining eslint things

volnei committed:
add tailwind directives

volnei committed:
Merge branch 'main' into feat/introduce-biome

volnei commented on biome.json:
```diff
@@ -0,0 +1,188 @@
+{
```
not related.

keithwillcode commented on apps/api/v2/package.json:
```diff
@@ -86,7 +86,7 @@
     "zod": "3.25.76"
   },
   "devDependencies": {
-    "@calcom/eslint-config": "workspace:*",
```
This makes me happy

keithwillcode commented on apps/api/package.json:
```diff
@@ -13,8 +13,5 @@
     "http": "0.0.1-security",
     "http-proxy-middleware": "2.0.9"
-  },
-  "devDependencies": {
```
This makes me happy

keithwillcode commented on packages/ui/components/icon/icon-names.ts:
```diff
@@ -1,159 +1,159 @@
-  | "webhook"
-  | "x"
-  | "zap";
+	| "activity"
```
@volnei Is this spacing correct?

volnei replied:
will fix in a follow up with other fixes!

keithwillcode reviewed (approved):
Spoke with @volnei about the fact that we have
Found 21 errors.
Found 49 warnings.

when running lint. There will be follow up PRs to fix these ASAP.
